dart/dynamics/BodyNode.cpp

```diff
@@ -896,6 +819,60 @@ const Joint* BodyNode::getChildJoint(size_t _index) const
 }
 
 //==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape,
+                                     const std::string& name)
+{
+  ShapeNode::Properties properties;
+  properties.mShape = shape;
+
+  return createNode<ShapeNode>(properties);
+}
+
+//==============================================================================
+ShapeNode* BodyNode::createShapeNode(const ShapePtr& shape, const char* name)
+{
+  return createShapeNode(shape, std::string(name));
+}
+
+//==============================================================================
+size_t BodyNode::getNumShapeNodes() const
+{
+  return getNumNodes<ShapeNode>();
+}
+
+//==============================================================================
+ShapeNode* BodyNode::getShapeNode(size_t index)
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+const ShapeNode* BodyNode::getShapeNode(size_t index) const
+{
+  return getNode<ShapeNode>(index);
+}
+
+//==============================================================================
+std::vector<ShapeNode*> BodyNode::getShapeNodes()
+{
+  std::vector<ShapeNode*> shapeNodes(getNumShapeNodes());
+
+  auto numShapeNode = getNumShapeNodes();
+  for (auto i = 0u; i < numShapeNode; ++i)
+    shapeNodes[i] = getShapeNode(i);
+
+  return shapeNodes;
+}
```

Comment:
> I thought returning the vector by value would create the vector not only the first time but also whenever it gets called. Is it incorrect?

If it returns by value, it will potentially have to allocate memory once each time it gets called, unless it can avoid that by using return value optimization. But one memory allocation is the worst possible case thanks to move semantics.

The danger of this static variable is that users are likely to catch the return of this function in a const-reference, but the contents of the vector that it's referring to would change as soon as the function gets called on a different ``BodyNode`` instance.